Code Review PR

aschmahmann opened the pull request:
This PR can be used to create some discussion on what's going on in the code, why it was done this way, and hopefully make it better.

aschmahmann committed:
Initial implementation (#1)

Initial implementation containing a WASM bencode codec and BitTorrent v1 ADL as well as wrappers for working with WASM codecs and ADLs from Go.

aschmahmann committed:
docs: remove duplicate MIT license

aschmahmann committed:
add comments to Go code

aschmahmann committed:
docs: add some comments to the Rust code

RangerMauve commented on wasmlib/helpers/src/lib.rs:
```diff
@@ -0,0 +1,407 @@
+Items still unclear:
+- What to do about 32 vs 64 bit? Do we just define everything as 32 bit until 64 bit wasm is everywhere?
+- How to free memory, both that used by the host to pass data in and for data returned (e.g. strings)
+- If/how we allow multiple modules to work together without going through the host
```
I think this is where the WRAP standard from #6 would come in handy

RangerMauve commented on wasmlib/helpers/src/lib.rs:
```diff
@@ -0,0 +1,407 @@
+// Extern Functions
+
+#[cfg(target_arch = "wasm32")]
+extern "C" {
```
It'd be nice if the C ABI we expect hosts to implemented could be put in a separate document to make it instantly visible.

aschmahmann replied:
Do you mean as text or as code? As code it's mostly in this file, and we can just move the few helper functions to a separate file from the type and function definitions.

RangerMauve commented on wasmlib/helpers/src/lib.rs:
```diff
@@ -0,0 +1,407 @@
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlMap {
+    unsafe fn new_map_iter(&self, adlptr: *const c_void) -> *mut c_void;
```
Could we have "seeking" be part of this? Or having some extra options that could be passed in.

aschmahmann replied:
Yep, I mostly copied these ADLs to match the ones from go-ipld-prime https://github.com/ipld/go-ipld-prime/blob/adf99676901e35d42bba09c67a236ee0c87675f0/datamodel/node.go#L60. If you've got to add totally game to add them.

I think being able to work with chunks of a map asynchronously instead of via an iterator is a good idea.

RangerMauve reviewed (commented):
Did you look at existing WASM ABIs when you were designing this one? If so could you link to them.

It'd be nice to have the "motivation" for the API design so we can talk about what tradeoffs are important and which have been made to create the current state.

aschmahmann commented on wasmlib/helpers/src/lib.rs:
```diff
@@ -0,0 +1,407 @@
+}
+
+/*
+/// Given a pointer to the start of a byte array of WAC data
```
This is mostly me being bad at Rust/WASM things.

What I want:
1. A way to specify the ABI
2. A way to make it easy for people to make new codecs/ADLs and just fill in the blanks, like fulfilling an interface contract
3. I want my code to compile

The `#[no_mangle]` public functions are part of the ABI, but I can't figure out how I'd include them in a new module since I don't know how to get the new module to make data available to the public function. Suggestions welcome, but I'd prefer not to have to make a WASM "init" call to set things up and IIUC that doesn't happen automatically like it might in some other environments.

So this is basically the code you can copy-paste to fulfill the ABI in your new WASM module. You fulfill the traits and then have the `#[no_mangle]` functions call out to your trait implementation.

Better ideas super duper welcome

aschmahmann committed:
chore: move to ipld org

Commit:
chore: Update .github/dependabot.yml [skip ci]

dependabot[bot] committed:
chore(deps): bump actions/setup-go from 2 to 4 (#11)

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 2 to 4.
- [Release notes](https://github.com/actions/setup-go/releases)
- [Commits](https://github.com/actions/setup-go/compare/v2...v4)

---
updated-dependencies:
- dependency-name: actions/setup-go
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump actions/checkout from 2 to 3 (#10)

Bumps [actions/checkout](https://github.com/actions/checkout) from 2 to 3.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump actions/cache from 1 to 3 (#9)

Bumps [actions/cache](https://github.com/actions/cache) from 1 to 3.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v1...v3)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
chore: Update .github/dependabot.yml [skip ci]

web3-bot committed:
ci: uci/copy-templates (#16)

* chore: add or force update .github/workflows/stale.yml

* chore: add or force update .github/workflows/generated-pr.yml
